File: LeetCodeProblemsCMake/Problems/LeetCode2929DistributeCandiesAmongChildrenII.cpp
//
// Created by saeed on 6/1/25.
//

#include "LeetCode2929DistributeCandiesAmongChildrenII.h"

#include <algorithm>
#include <iostream>
#include <ostream>
using namespace std;
long long LeetCode2929DistributeCandiesAmongChildrenII::distributeCandies(int n, int limit) {
    int minLim=std::min(n,limit);
    long long sum=0;
    long long stepCombinations=0;
    for(int i=0;i<=minLim;i++) {
        stepCombinations=std::max(std::min(n,i+minLim)-std::max(i,n-minLim)+1,0);
        sum+=stepCombinations;
    }
    return sum;
}

void LeetCode2929DistributeCandiesAmongChildrenII::Run() {
    int n=3;
    int limit=3;
    cout<<distributeCandies(n,limit)<<endl;
}

File: LeetCodeProblemsGo/LeetCode2929DistributeCandiesAmongChildrenII/LeetCode2929DistributeCandiesAmongChildrenII.go
package LeetCode2929DistributeCandiesAmongChildrenII

func distributeCandies(n int, limit int) int64 {
	minLimit := limit
	if n < limit {
		minLimit = n
	}
	var result int64 = 0
	var stepSum int64 = 0
	for i := 0; i <= minLimit; i++ {
		stepSum = int64(max(min(n, i+minLimit)-max(i, n-minLimit)+1, 0))
		result += stepSum
	}
	return result
}
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
